lib/services/fanout.go

```diff
@@ -354,3 +359,87 @@ func (w *fanoutWatcher) Error() error {
 		return nil
 	}
 }
+
+// fanoutSetSize is the number of members in a fanout set.  selected based on some experimentation with
+// the FanoutSetRegistration benchmark.  This value keeps 100K concurrent registrations well under 1s.
+const fanoutSetSize = 128
+
+// FanoutSet is a collection of separate Fanout instances. It exposes an identical API, and "load balances"
+// watcher registration across the enclosed instances. In very large clusters it is possible for tens of
+// thousands of nodes to simultaneously request watchers. This can cause serious contention issues. FanoutSet is
+// a simple but effective solution to that problem.
+type FanoutSet struct {
+	// rw mutex is used to ensure that Close and Reset operations are exclusive,
+	// since these operations close watchers. Enforcing this property isn't strictly
+	// necessary, but it prevents a scenario where watchers might observe a reset/close,
+	// attempt re-registration, and observe the *same* reset/close again. This isn't
+	// necessarily a problem, but it might confuse attempts to debug other event-system
+	// issues, so we choose to avoid it.
+	rw      *sync.RWMutex
+	counter *atomic.Uint64
+	members []*Fanout
+}
+
+// NewFanoutSet creates a new FanoutSet instance in an uninitialized
+// state.  Until initialized, watchers will be queued but no
+// events will be sent.
+func NewFanoutSet() FanoutSet {
+	members := make([]*Fanout, 0, fanoutSetSize)
```

Comment:
nit: Let length = capacity, we do that anyway below.

```suggestion
	members := make([]*Fanout, fanoutSetSize)
```

or, if you took the suggestion above, then this is all we need:

```suggestion
	members := make([]Fanout, fanoutSetSize)
```

Reply:
> we do that anyway below

I'm not sure I follow.  If we set length to `fanoutSetSize` then the slice gets filled with zeroed fanouts, and the below loop appends an additional `fanoutSetSize` worth of nonzero fanouts to that slice.

I could change the below loop to use index based assignment instead of appending.  Is there some advantage to doing it that way?  I've never really looked into the substantive difference.

Reply:
I prefer doing the index-based assignments when I know the final length of the array, but I don't suppose it's much of a practical difference (hence the nit tag in the comment).

Feel free to push back / keep as-is.